src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs

```diff
@@ -1125,4 +1127,95 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
             codeStream.Emit(ILOpcode.call, emitter.NewToken(helper));
         }
     }
+
+    class CustomTypeMarshaller : Marshaller
+    {
+        protected override void TransformManagedToNative(ILCodeStream codeStream)
+        {
+            var customMarshallerType = Context.SystemModule.GetKnownType("System.Runtime.InteropServices", "ICustomMarshaler");
+            InitializeMarshaller(customMarshallerType, out ILLocalVariable lMarshaller);
+
+            var manageToNativeMethod = customMarshallerType.GetKnownMethod(
+                "MarshalManagedToNative",
+                new MethodSignature(MethodSignatureFlags.None, 0, Context.GetWellKnownType(WellKnownType.IntPtr), new[] { Context.GetWellKnownType(WellKnownType.Object) }));
+
+            ILEmitter emitter = _ilCodeStreams.Emitter;
+            codeStream.EmitLdLoc(lMarshaller);
+            LoadManagedValue(codeStream);
+            codeStream.Emit(ILOpcode.callvirt, emitter.NewToken(manageToNativeMethod));
+            StoreNativeValue(codeStream);
+
+            // Call CleanUpNativeData on cleanup code stream.
+            var cleanupNativeDataMethod = customMarshallerType.GetKnownMethod(
+                "CleanUpNativeData",
+                new MethodSignature(MethodSignatureFlags.None, 0, Context.GetWellKnownType(WellKnownType.Void), new[] { Context.GetWellKnownType(WellKnownType.IntPtr) }));
+
+            _ilCodeStreams.CleanupCodeStream.EmitLdLoc(lMarshaller);
+            LoadNativeValue(_ilCodeStreams.CleanupCodeStream);
+            _ilCodeStreams.CleanupCodeStream.Emit(ILOpcode.callvirt, emitter.NewToken(cleanupNativeDataMethod));
+        }
+
+        protected override void TransformNativeToManaged(ILCodeStream codeStream)
+        {
+            var customMarshallerType = Context.SystemModule.GetKnownType("System.Runtime.InteropServices", "ICustomMarshaler");
+            InitializeMarshaller(customMarshallerType, out ILLocalVariable lMarshaller);
+
+            var marshalNativeToManagedMethod = customMarshallerType.GetKnownMethod(
+                "MarshalNativeToManaged",
+                new MethodSignature(MethodSignatureFlags.None, 0, Context.GetWellKnownType(WellKnownType.Object), new[] { Context.GetWellKnownType(WellKnownType.IntPtr) }));
+
+            ILEmitter emitter = _ilCodeStreams.Emitter;
+            codeStream.EmitLdLoc(lMarshaller);
+            LoadNativeValue(codeStream);
+            codeStream.Emit(ILOpcode.callvirt, emitter.NewToken(marshalNativeToManagedMethod));
+            StoreManagedValue(codeStream);
```

Comment:
Do you need to call `ICustomMarshaler.CleanUpManagedData` here?

Reply:
Honestly, if not Michal help, I would never found these places. Even when I'm looking from them, I cannot find who emit calls to `CleanUpManagedData` so I cannot follow existing logic and rely only on my gut feling and tests.

If you think, that I should place that in `CleanUpManagedData` in `Marshaller.EmitCleanupManaged`, then I would 
- have to exclude call that for `TransformNativeToManaged` or it will fails.
- I would not have easy access to `lMarshaller` variable. This is not so problematic, just convinient.

Or did you mean that I select incorrect code stream?

Reply:
The managed to native, and native to managed directions are symmetric. 

Managed to native should calls `ICustomMarshaler.MarshalManagedToNative` and `ICustomMarshaler.CleanUpNativeData`. The code in this PR seems to be doing it correctly.

Native to managed should call `ICustomMarshaler.MarshalNativeToManaged` and  `ICustomMarshaler.CleanUpManagedData`.  The code in this PR seems to be missing the `CleanUpManagedData` part.

Reply:
Yeah. I agree. Still I did not find original code, maybe I should look more. And that case did not covered by tests.

https://github.com/dotnet/runtimelab/blob/1e626fccf9cd223502a3c5fc211e4f7a91c03591/src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs#L141

I suspect that `ICustomMarshaler.CleanUpManagedData` probably related to delegates marshalling, which currently did not tested. I can go to runtime repo and add additional tests which highlight that behaviour. Then my life would be slightly easier, since I may rely on `runtime` tests, and your experience.

Reply:
The CoreCLR code that calls the cleanup method is: https://github.com/dotnet/runtime/search?q=InvokeCleanUpManagedMeth

I am not able to find any tests for custom marshaler and delegates. It would be nice to add a few.

Reply:
https://github.com/dotnet/runtime/pull/62691 this is what I plan to add. I hope this would be enough.